trie/ctrie/ctrie.go

```diff
@@ -258,21 +305,52 @@ func (c *Ctrie) Lookup(key []byte) (interface{}, bool) {
 // Remove deletes the value for the associated key, returning true if it was
 // removed or false if the entry doesn't exist.
 func (c *Ctrie) Remove(key []byte) (interface{}, bool) {
+	c.assertReadWrite()
 	return c.remove(&entry{key: key, hash: c.hash(key)})
 }
 
+// Snapshot returns a stable, point-in-time snapshot of the Ctrie.
+func (c *Ctrie) Snapshot() *Ctrie {
+	for {
+		root := c.readRoot()
+		main := gcasRead(root, c)
+		if c.rdcssRoot(root, main, root.copyToGen(&generation{}, c)) {
+			return newCtrie(root.copyToGen(&generation{}, c), c.hashFactory, c.readOnly)
+		}
+	}
+}
+
+// ReadOnlySnapshot returns a stable, point-in-time snapshot of the Ctrie which
+// is read-only. Write operations on a read-only snapshot will panic.
+func (c *Ctrie) ReadOnlySnapshot() *Ctrie {
+	if c.readOnly {
+		return c
+	}
+	for {
+		root := c.readRoot()
+		main := gcasRead(root, c)
+		if c.rdcssRoot(root, main, root.copyToGen(&generation{}, c)) {
+			return newCtrie(root, c.hashFactory, true)
+		}
+	}
+}
+
+func (c *Ctrie) assertReadWrite() {
+	if c.readOnly {
+		panic("Cannot modify read-only snapshot")
```

Comment:
K. Fair enough. The more I've been writing Go the more I've hit these kind of errors. Not really ideal to bubble up (or at times stuck in go routines so not easily done without specific channels) but also maybe not "panic" level.